text/3729-sized-hierarchy.md

```diff
@@ -0,0 +1,1412 @@
+a `const ValueSized` bound.
+
+**Edition change:** In the current edition,`?Sized` will be syntatic sugar for
+a `const ValueSized` bound. In the next edition, use of `?Sized` syntax will be prohibited
+over an edition and all uses of it will be rewritten to a `const ValueSized` bound.
+
+A default implicit bound of `const Sized` is added by the compiler to every type
+parameter `T` that does not have an explicit `Sized`, `?Sized`, `const ValueSized`,
+`ValueSized` or `Pointee` bound. It is backwards compatible to change
+the current implicit `Sized` bound to an `const Sized` bound as every type which
+exists currently will implement `const Sized`.
+
+**Edition change:** In the current edition, all existing `Sized` bounds will be
+syntatic sugar for `const Sized` (both explicitly written and implicit). In the next
+edition, existing `Sized` bounds will be rewritten to `const Sized` bounds and new
+bare `Sized` bounds will be non-const. `Sized` supertraits (such as that on `Clone`
+or `Default`) will not be sugar for `const Sized`, these will remain bare `Sized`.
+If traits with a `Sized` supertrait are later made const, then their supertrait
+would be made `~const Sized`.
+
+An implicit `const ValueSized` bound is added to the `Self` type of traits. Like
+implicit `const Sized` bounds, this is omitted if an explicit `const Sized`, `Sized`,
+`ValueSized` or `Pointee` bound is present.
+
+As `ValueSized` and `Pointee` are not default bounds, there is no equivalent to `?Sized`
+for these traits.
+
+## `size_of` and `size_of_val`
+[size-of-and-size-of-val]: #size-of-and-size-of-val
+
+Runtime-sized types should be able to be passed to both `size_of` and `size_of_val`,
+but only at runtime, which requires ensuring these functions are only const if their
+arguments have const implementations of the relevant sizedness traits.
+
+[`size_of`][api_size_of] is a const-stable function since Rust 1.24.0 and currently
+accepts a `T: Sized` bound. Therefore, even when used in a const context, `size_of`
+could accept a runtime-sized type. It is therefore necessary to modify the bounds of
+`size_of` to accept a `T: ~const Sized`, so that `size_of` is a const function
+if-and-only-if `Sized` has a `const` implementation.
+
+```rust=
+pub const fn size_of<T: ~const Sized>() -> usize {
+    /* .. */
+}
+```
+
+This has the potential to break existing code like `uses_size_of` in the below
+example. However, due to the changes described in [`Sized` bounds][sized-bounds]
+(changing the implicit `T: Sized` to `T: const Sized`, and treating explicit
+`T: Sized` bounds as `T: const Sized` in the current edition), this code would
+not break:
+
+```rust=
+fn uses_size_of<T: Sized>() -> usize {
+    const { std::mem::size_of<T>() }
+}
+
+fn another_use_of_size_of<T: Sized>() -> [u8; size_of::<T>()] {
+    std::array::repeat(0)
+}
+```
+
+[`size_of_val`][api_size_of_val] is also const-stable, so like `size_of` above,
```

Comment:
Not yet. FCP is still ongoing in https://github.com/rust-lang/rust/issues/46571. We can halt that if we have to.

Reply:
Yeah, I wrote this assuming that FCP goes ahead.